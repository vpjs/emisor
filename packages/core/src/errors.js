export class EmisorTypeError extends TypeError {
  constructor(param, expect, given) {
    if (!Array.isArray(expect)) {
      expect = [expect];
    }
    super(`${param} has to be a ${expect.join(' or ')}, ${typeof given} given`);
  }
}

export class EmisorError extends Error {}

/**
 * Error object
 */
export class EmisorPluginError extends EmisorError {}


export class EmisorPluginTypeError extends EmisorTypeError {}