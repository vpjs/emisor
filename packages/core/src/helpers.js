function getName (obj) {
  return ({}).toString.call(obj);
}

/**
 * is given value a symbol
 * @param {*} value
 * @returns {boolean}  
 */
export function isSymbol (value) {
  return typeof value === 'symbol' || getName(value) == '[object Symbol]';
}

/**
 * is given value a RegExp
 * @param {*} value
 * @returns {boolean}  
 */
export function isRegExp (value) {
  return getName(value) == '[object RegExp]';
}
/**
 * check of given value is a string
 * @param {*} value
 * @returns {boolean} 
 */
export function isString (value) {
  return typeof value === 'string' || getName(value) == '[object String]';
}

/**
 * check of given value is a function
 * @param {*} value
 * @returns {boolean} 
 */
export function isFunction (value) {
  return typeof value === 'function';
}