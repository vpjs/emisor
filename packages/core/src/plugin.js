/**
 * @typedef {import('./hook').EmisorHook} EmisorHook
 * @typedef {import('./hook').EmisorHookEventStr} EmisorHookEventStr
 */

/**
 * @typedef {object} EmisorPluginHook 
 * @prop {EmisorHook["pluginApi"]} beforeOn
 * @prop {EmisorHook["pluginApi"]} afterOn
 * @prop {EmisorHook["pluginApi"]} beforeEmit
 * @prop {EmisorHook["pluginApi"]} afterEmit
 * @prop {EmisorHook["pluginApi"]} beforeOff
 * @prop {EmisorHook["pluginApi"]} afterOff
 * @prop {EmisorHookEventStr["pluginApi"]} eventStr
 */

/**
 * @typedef {object} IEmisorPlugin
 * @prop {(EmisorPluginHook)=>void} install - install plugin
 */

/**
 * Error object
 */
export class EmisorPluginError extends Error {}

export class EmisorPluginTypeError extends EmisorPluginError {
  constructor(param, expect, given) {
    super(`${param} has to be a ${expect}, ${typeof given} given`);
  }
}

/**
 * @implements {IEmisorPlugin}
 */
export class EmisorPlugin {
  install () {
    throw new EmisorPluginError('Plugin is missing a install method');
  }
}