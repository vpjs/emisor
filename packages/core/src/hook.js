import { EmisorPluginTypeError } from './plugin';
import { isSymbol, isString, isRegExp, isFunction } from './helpers';

/** @typedef {import('./').EmisorEvent} EmisorEvent */
/** @typedef {import('./').EmisorEventObject} EmisorEventObject */

/**
 * @typedef {object} EmisorHookMeta
 * @prop {EmisorEventObject} meta.event
 * @prop {*} [meta.payload]
 * @prop {object} meta.storage
 * @prop {*} [meta.options]
 */

/**
 * @callback EmisorHookCallback
 * @param {EmisorHookMeta} meta
 * @param {import('./').EmisorHookAPI} Emisor
 * @returns {object}
 */

/**
 * @callback EmisorHookFnc
 * @param {EmisorEventObject} event
 * @param {*} [payload]
 * @return {EmisorHookCallback}
 */


export class EmisorHook {
  /**
   * @type {Map<string|Symbol, EmisorHookCallback>}
   */
  #keyCallbacks = new Map();

  /**
   * @type {Set<EmisorHookCallback>}
   */
  #allCallbacks = new Set();



  get pluginApi () {
    return {
      /**
       * Register a hook based on a key
       * @param {string|Symbol} key
       * @param {EmisorHookCallback} callback 
       */
      key: (key, callback) => {
        if (!isString(key) && !isSymbol(key)) {
          throw new EmisorPluginTypeError('key', 'string or Symbol', key);
        }
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#keyCallbacks.set(key, callback);
      },
      /**
       * Register a hook all events
       * @param {EmisorHookCallback} callback 
       */
      all: (callback) => {
        this.#allCallbacks.add(callback);
      } 
    };
  }


  /**
   * 
   * @param {import('./').EmisorHookAPI} Emisor
   * @returns {EmisorHookFnc[]}
   */
  getAllHooks (Emisor) {
    return Array.from(this.#allCallbacks)
      .map((callback) => {
        let  storage = {};
        return (event, payload) => callback({
          event,
          payload,
          storage
        }, Emisor);
      });
  }

  /**
   * @param {object} options 
   * @param {import('./').EmisorHookAPI} Emisor
   * @returns {EmisorHookFnc[]}
   */
  getOptionHooks (options, Emisor) {
    return Object.entries(options)
      .filter(([key]) => this.#keyCallbacks.has(key))
      .map(([key, $options]) => {
        let storage = {};
        /**
         * @param {import('./').EmisorEvent} event
         * @param {*} [payload]
         * @returns {EmisorHookCallBack}
         */
        return (event, payload) => this.#keyCallbacks.get(key)({
          storage,
          payload,
          options: $options,
          event
        }, Emisor);
      });
  }
}

/**
 * @callback EmisorHookEventStrCallBack
 * @param {string} $event
 * @return {Object<string|Symbol, any>}
 */

export class EmisorHookEventStr {
  /**
   * @type {string}
   */
  #postfixSeparator

  /**
   * @type {Map<RegExp, EmisorHookEventStrCallBack>}
   */
  #postfixCallbacks = new Map();

  /**
   * @param {string} s
   */
  set postfixSeparator(s) {
    if (isString(s) && s.length === 1) {
      this.#postfixSeparator = s;
    }
  }

  get pluginApi () {
    return {
      /**
       * Register a postfix handler
       * @param {RegExp} regex
       * @param {EmisorHookEventStrCallBack} callback
       */
      postfix: (regex, callback) => {
        if (!isRegExp(regex)) {
          throw new EmisorPluginTypeError('regex', 'RegExp', regex);
        }
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#postfixCallbacks.set(regex, callback);
      }
    };
  }

  /**
   * 
   * @param {string} event 
   * @return {{event:string, options: Object<string|Symbol, any>}}
   */
  parseStr (event) {
    return this.#postfixStr(event);
  }

  /**
   * 
   * @param {string} rawEvent 
   * @return {{event:string, options: Object<string|Symbol, any>}}
   */
  #postfixStr (rawEvent) {
    let options = {},
        [event, postfix] = rawEvent.split(this.#postfixSeparator);
    //only when event has a postfix
    if (postfix) {
      this.#postfixCallbacks.forEach((hook, reg) => {
        if (reg.test(postfix)) {
          options = {
            ...options,
            ...hook(postfix)
          };
        }
      });
    }
    return {
      event,
      options
    }
  }
}