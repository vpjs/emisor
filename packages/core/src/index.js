import { isString, isSymbol, isFunction } from './helpers';
import { EmisorHook, EmisorHookEventStr } from './hook';
export * as helpers from './helpers';
export * from './plugin';

/**
 * @typedef {import('./hook').EmisorHookFnc} EmisorHookFnc
 */
/**
 * @typedef {import('./hook').EmisorHookMeta} EmisorHookMeta
 */
/**
 * @typedef {import('./plugin').IEmisorPlugin} IEmisorPlugin
 */

/**
 * @typedef {object} EmisorOptions
 * @prop {string} [nsSeparator='.'] namespace separator options
 * @prop {string} [postfixSeparator=':'] postfix separator
 * @prop {IEmisorPlugin[]} [plugins] plugins
 */
 
/**
 * @typedef {string|Symbol} EmisorEvent
 */

/**
  * @typedef {Object<string|number, any>} EmisorEventOptions
  */

 
/**
 * @typedef {Map<string, EmisorHookFnc[]>} EmisorPluginHooksMap
 */

/**
 * @typedef {Map<EmisorEventHandler, EmisorPluginHooksMap>} EmisorSubsMap
 */

/**
 * @typedef {object} EmisorEventObject
 * @prop {string} [id] of event `1-1`
 * @prop {number} time unix timestamp of publish time
 * @prop {EmisorEvent} event that triggered the handler
 * @prop {EmisorEventHandler} [handler]
 * @prop {*} [tag]
 */


/**
  * @typedef {object} EmisorChainingAPI
  * @prop {EmisorCore["on"]} on
  * @prop {EmisorCore["emit"]} emit
  * @prop {EmisorCore["off"]} off
  */

/**
  * @typedef {object} EmisorHookAPI
  * @prop {EmisorCore["on"]} on
  * @prop {EmisorCore["emit"]} emit
  * @prop {EmisorCore["off"]} off
  * @prop {Function} rawEmit
  * @prop {Function} parseEvent
  */

/**
 * @callback EmisorEventHandler
 * @param {*} data
 * @param {EmisorEventObject} $event
 */

const DEFAULT_NS_SEPARATOR = '.';
const DEFAULT_POSTFIX_SEPARATOR = ':';


export default class EmisorCore {
  /**
   * @type {Map<EmisorEvent, EmisorSubsMap>}
   */
  #subs = new Map()

  /**
   * @type {string} namespace separator
   */
  #nsSeparator
  /**
   * Hook that will be trigger before subscribed to a event 
   */
  #beforeOnHook = new EmisorHook()
  /**
   * Hook that will be trigger after subscribed to a event 
   */
  #afterOnHook = new EmisorHook()
  /**
   * Hook that will be trigger before event emit
   */
  #beforeEmitHook = new EmisorHook()
  /**
   * Hook that will be trigger after event emit
   */
  #afterEmitHook = new EmisorHook()
  /**
   * Hook that will be trigger before event is unsubscribed
   */
  #beforeOffHook = new EmisorHook()
  /**
   * Hook that will be trigger after event is unsubscribed
   */
  #afterOffHook = new EmisorHook()

  #eventStrHook = new EmisorHookEventStr()

  /**
   * @type {EmisorHookAPI}
   */
  get #hookApi () {
    return {
      on: (...args) => this.on(...args),
      emit: (...args) => this.emit(...args),
      off: (...args) => this.off(...args),
      rawEmit: (...args) => this.#rawEmit(...args),
      parseEvent: (event) => this.#parseEvent(event)
    }
  }

  /**
   * 
   * @param {EmisorOptions} options 
   */
  constructor ({
    nsSeparator = DEFAULT_NS_SEPARATOR,
    postfixSeparator = DEFAULT_POSTFIX_SEPARATOR,
    plugins = []
  } = {}) {
    this.#nsSeparator = nsSeparator;
    this.#eventStrHook.postfixSeparator = postfixSeparator;

    plugins.forEach((plugin) => {
      plugin.install({
        beforeOn: this.#beforeOnHook.pluginApi,
        afterOn: this.#afterOnHook.pluginApi,
        beforeEmit: this.#beforeEmitHook.pluginApi,
        afterEmit: this.#afterEmitHook.pluginApi,
        beforeOff: this.#beforeOffHook.pluginApi,
        afterOff: this.#afterOffHook.pluginApi,
        eventStr: this.#eventStrHook.pluginApi
      });
    });
  }

  /**
   * Subscribe to a event
   * @param {EmisorEvent} event 
   * @param {EmisorEventHandler} handler 
   * @param {object} [options]
   */
  on (event, handler, options = {}) {
      

    if(isString(event)) {
      let result = this.#eventStrHook.parseStr(/** @type {string}*/ (event));
      event = result.event;
      options = {
        ...result.options,
        ...options
      };
    }

    if(!this.#subs.has(event)) {
      this.#subs.set(event, new Map());
    }
    let beforeOn = [
        ...this.#beforeOnHook.getOptionHooks(options, this.#hookApi),
        ...this.#beforeOnHook.getAllHooks(this.#hookApi)
      ],
      afterOn = [
        ...this.#afterOnHook.getOptionHooks(options, this.#hookApi),
        ...this.#afterOnHook.getAllHooks(this.#hookApi)
      ],
      $event = {
        time: Date.now(),
        event,
        handler
      };
    
    //before on hooks
    beforeOn.forEach((hook) => {
      let {handler: overwriteHandler} = hook($event) || {};
      if (overwriteHandler) {
        handler = overwriteHandler;
      }
    });
    
    //subscribe
    this.#subs.get(event).set(handler, new Map([
      ['beforeEmit', this.#beforeEmitHook.getOptionHooks(options, this.#hookApi)],
      ['afterEmit', this.#afterEmitHook.getOptionHooks(options, this.#hookApi)],
      ['beforeOff', this.#beforeOffHook.getOptionHooks(options, this.#hookApi)],
      ['afterOff', this.#afterOffHook.getOptionHooks(options, this.#hookApi)]
    ]));
    //after on hooks
    afterOn.forEach((hook) => hook($event));

    return this;
  }
  /**
   * Unsubscribe
   * If no `event` is given, all subscribe will be unsubscribed 
   * if no `handler` is given all subscribe of the given `event` will be unsubscribed
   * @param {EmisorEvent} [event] unsubscribe to a specific event
   * @param {EmisorEventHandler} [handler] unsubscribe to a specific handler
   */
  off (event, handler) {
    //remove all
    if (!event) {
      this.#subs = new Map();
      return this;
    }
    let handlers = this.#subs.get(event);
    //if a event doesn't have any handlers we can don't unsubscribe anything 
    if(!handlers) {
      return this;
    }
    //remove all handlers of event
    if (!handler) {
      handlers.clear();
    } else {
      handlers.delete(handler);
    }
    return this;
  }

  /**
   * 
   * @param {EmisorEvent} event 
   * @param {*} [payload] 
   * @return {EmisorChainingAPI}
   */
  emit (event, payload) {
    this.#rawEmit(event, payload);
    return this;
  }

  /**
   * @param {EmisorEvent} event
   * @returns {EmisorEvent[]} 
   */
  #parseEvent (event) {
    if (isSymbol(event)) {
      return [event, '*'];
    } else if (isString(event)) { 
      let ns = /** @type {string} **/ (event).split(this.#nsSeparator);
      return [
        event,
        `${event}${this.#nsSeparator}*`,
        ...ns.map((_, i, arr) => {
          let x = [...arr];
          x[i] = '*';
          return [
            x.join(this.#nsSeparator),
            [...arr.slice(0,i), '*'].join(this.#nsSeparator)
          ]
        }).flat()
      ].filter((v, i, a) => a.indexOf(v) === i)
    }
  }

  /**
   * 
   * @param {EmisorEvent} event 
   * @param {*} [payload]
   * @param {EmisorEventHandler|EmisorEventHandler[]} [only]
   * @param {*} [tag]
   */
  async #rawEmit (event, payload, only, tag) {
    /** @type {Array<[EmisorEventHandler, EmisorPluginHooksMap]>} */
    let subs = [], 
      parsedEvents = this.#parseEvent(event);
    only = /** @type {EmisorEventHandler[]} **/ (isFunction(only) ? [only] : only);
    parsedEvents.forEach((event) => {
      if(this.#subs.has(event)) {
        subs = [
          ...subs,
          ...Array.from(this.#subs.get(event))
          .filter(([handler]) => only ? /** @type {EmisorEventHandler[]} **/ (only).includes(handler) : true)
        ];
      }
    });

    if (subs.length === 0) {
      this.#beforeEmitHook.getAllHooks(this.#hookApi)
      .forEach((hook) => hook({
        event,
        time: Date.now(),
        tag
      }, payload));
    }
    subs.map(([handler, hooks]) => Promise.resolve({
      event,
      tag,
      handler,
      payload,
      hooks
    }))
    .forEach((p) => p.then((d) => this.#publish(d)));
  }

  /**
   * @param {{event: EmisorEvent, handler: EmisorEventHandler, payload: any, tag: any, hooks: EmisorPluginHooksMap}} param1
   */
  #publish ({event, handler, payload, hooks, tag}) {
    let $event = {
      event,
      handler,
      time: Date.now(),
      id: '1',
      tag
    };
    //before emit hooks
    [
      ...hooks.get('beforeEmit'),
      ...this.#beforeEmitHook.getAllHooks(this.#hookApi)
    ].forEach((hook) => hook($event, payload));
    //emit
    handler(payload, $event);
    //after emit hooks
    [
      ...hooks.get('afterEmit'),
      ...this.#afterEmitHook.getAllHooks(this.#hookApi)
    ].forEach((hook) => hook($event, payload));
  }
}

export {
  EmisorCore
}