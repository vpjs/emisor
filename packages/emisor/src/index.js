import { EmisorCore } from '@emisor/core';
import { EmisorPluginCount } from '@emisor/plugin-count';
import { EmisorPluginHistory, MODE_DEFAULT_ALLOW, MODE_DEFAULT_DENY } from '@emisor/plugin-history';

/** @typedef {import('@emisor/core').EmisorEvent} EmisorEvent */
/** @typedef {import('@emisor/core').EmisorEventHandler} EmisorEventHandler */


//history modes
export const HISTORY_MODE_DEFAULT_ALLOW = MODE_DEFAULT_ALLOW;
export const HISTORY_MODE_DEFAULT_DENY = MODE_DEFAULT_DENY;

export class Emisor extends EmisorCore {
  #historyKey;
  #countKey;
  
  /**
   * @param {import('@emisor/core').EmisorOptions} options
   * @param {object} [pluginConfig]
   */
  constructor ({
    plugins = [], 
    ...options
  } = {}, {
    history = {}, 
    count = {}
  } = {}) {
    super({
      plugins: [
        new EmisorPluginCount(count),
        new EmisorPluginHistory(history),
        ...plugins
      ],
      ...options,
      chain: [
        ...(options?.addToChain || []),
        'once',
        'many',
        'history',
        'historyOnce'
      ],
    });
    this.#historyKey = history.key || 'history';
    this.#countKey = count.key || 'count';
  }

  /**
   * @param {EmisorEvent} event
   * @param {EmisorEventHandler} handler
   */
  once (event, handler) {
    return this.on(event, handler, {
      [this.#countKey]: 1
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {number} count
   * @param {EmisorEventHandler} handler
   */
  many (event, count, handler) {
    return this.on(event, handler, {
      [this.#countKey]: count
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {number|boolean|EmisorEventHandler} length
   * @param {EmisorEventHandler} handler
   */
  history (event, length, handler) {
    if (typeof length === 'function') {
      handler = length;
      length = true;
    }
    return this.on(event, handler, {
      [this.#historyKey]: length
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {EmisorEventHandler} handler
   */
  historyOnce (event, handler) {
    return this.on(event, handler, {
      [this.#historyKey]: true ,
      [this.#countKey]: 1
    });
  }
}