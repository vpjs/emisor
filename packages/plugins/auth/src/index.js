import { EmisorPlugin, helpers } from '@emisor/core';

/**
 * 
 */
export class EmisorPluginAuth extends EmisorPlugin {


  /**
   */
  constructor () {
    super();
  }

  /**
  * @param {import('@emisor/core').EmisorPluginHooks} hook
  */
  install ({}) {
    
  }
}