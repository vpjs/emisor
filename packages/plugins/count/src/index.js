import { EmisorPlugin } from '@emisor/core';

/**
 * @typedef {import("@emisor/core").IEmisorPlugin} IEmisorPlugin
 */

const COUNT = Symbol();
const DEFAULT_KEY = 'count';
const POSTFIX_ON_KEY = /#(?<count>[1-9]\d*)/;

/**
 * @implements {IEmisorPlugin}
 */
export class EmisorPluginCount extends EmisorPlugin {
  #key

  /**
   * @param {object} [options]
   * @param {string} [options.key='count']
   */
  constructor ({key = DEFAULT_KEY} = {}) { 
    super();
    this.#key = key;
  }

  /**
   * @param {import('@emisor/core').EmisorPluginHook} hook 
   */
  install(hook) {
    hook.eventStr.postfix(POSTFIX_ON_KEY, ($event) =>  {
      let {count} = POSTFIX_ON_KEY.exec($event).groups;
      return {count};
    });
    hook.beforeOn.key(
      this.#key,
      (...args) => this.#beforeOn(...args)
    );
  }

  /**
   * 
   * @param {import('@emisor/core').EmisorHookMeta} meta
   * @param {import('@emisor/core').EmisorHookAPI} Emisor 
   */
  #beforeOn ({event: {handler}, options}, Emisor) {
    if (options > 0) {
      let count = options;
      return {
        /**
         * @type {import('@emisor/core').EmisorEventHandler}
         */
        handler: (data, $event) => {
          if (count === 1) {
            Emisor.off($event.event, $event.handler);
          }
          if (count) {
            handler(data, $event);
            count--;
          }
        }
      }
    }
  }
}