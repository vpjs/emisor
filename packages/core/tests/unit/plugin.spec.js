import { EmisorPlugin, EmisorPluginTypeError } from '../../src/plugin';

describe ('EmisorPlugin', () => {
  test('plugin should have method install', () => {
    expect(() => {
      new EmisorPlugin().install();
    }).toThrowErrorMatchingSnapshot();
  });

  test('plugin type error', ()  => {
    let error = new EmisorPluginTypeError('x','string', undefined);
    expect(error.message).toEqual('x has to be a string, undefined given');
  });
});