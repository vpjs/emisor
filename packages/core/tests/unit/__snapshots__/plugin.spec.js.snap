// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmisorPlugin plugin should have method install 1`] = `"Plugin is missing a install method"`;
